### DISCUSSION TITLE
Audio problem on Linux.

### DISCUSSION THREAD

**Discussion by Tresquel** [ppy/osu#15578] [Q&A] [answered]
Hello! I have a problem with sound inside of osu! which is that instead of having 2 output devices (laptop speakers and bluetooth headphones) I have a lot:
![image](https://user-images.githubusercontent.com/43729305/141325171-d6571660-f7d2-4bd7-b15d-183a0ba07981.png)
Now that wouldn't be a problem since I can pick from them and just see which is which but only few work but only work on my speakers,
the rest crashes bass (the library that you're using for sound). 

neofetch (maybe it'll be useful):
![image](https://user-images.githubusercontent.com/43729305/141325682-8c1114ec-29a8-4f55-8212-584d2cd1193f.png)

**Comment by peppy** [comment, marked as answer]
This is as you might expect an issue with your linux system configuration. Please read up on how to disable them yourself.

Please also attach your logs for one of the crashes so we can confirm it's not something we can handle more gracefully.

**Reply by Tresquel** [reply]
I ran `aplay -l` to list all my audio outputs but I only got my normal one and a few HDMI ones so I don't know how osu! is detecting more:

`aplay -l` output:
```
$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: PCH [HDA Intel PCH], device 0: ALC255 Analog [ALC255 Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 9: HDMI 3 [HDMI 3]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 10: HDMI 4 [HDMI 4]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```


As for the logs I don't know where they're located.


Edit:
When I was going thru the outputs to see which ones crash I found one output which sends audio to my headphones instead to my speakers.

**Reply by nahkd123** [reply]
> As for the logs I don't know where they're located.
> 

They're located in ``~/.local/share/osu/logs``

**Reply by smoogipoo** [reply]
I have the same thing, as I believe all/most linux users do. Never thought much about it because selecting one of these other devices does not crash the game - only outputs failure from bass.

@Tresquel please confirm that this doesn't crash the game for you either.

I think the issue is BASS listing the ALSA PCMs:
```
$ aplay --list-pcms

null
    Discard all samples (playback) or generate zero samples (capture)
samplerate
    Rate Converter Plugin Using Samplerate Library
speexrate
    Rate Converter Plugin Using Speex Resampler
jack
    JACK Audio Connection Kit
oss
    Open Sound System
pulse
    PulseAudio Sound Server
speex
    Plugin using Speex DSP (resample, agc, denoise, echo, dereverb)
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
default
    Default ALSA Output (currently PulseAudio Sound Server)
hdmi:CARD=NVidia,DEV=0
    HDA NVidia, HDMI 0
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=1
    HDA NVidia, HDMI 1
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=2
    HDA NVidia, HDMI 2
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=3
    HDA NVidia, HDMI 3
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=4
    HDA NVidia, HDMI 4
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=5
    HDA NVidia, HDMI 5
    HDMI Audio Output
usbstream:CARD=NVidia
    HDA NVidia
    USB Stream Output
hdmi:CARD=HDMI,DEV=0
    HDA ATI HDMI, HDMI 0
    HDMI Audio Output
hdmi:CARD=HDMI,DEV=1
    HDA ATI HDMI, HDMI 1
    HDMI Audio Output
hdmi:CARD=HDMI,DEV=2
    HDA ATI HDMI, HDMI 2
    HDMI Audio Output
hdmi:CARD=HDMI,DEV=3
    HDA ATI HDMI, HDMI 3
    HDMI Audio Output
hdmi:CARD=HDMI,DEV=4
    HDA ATI HDMI, HDMI 4
    HDMI Audio Output
hdmi:CARD=HDMI,DEV=5
    HDA ATI HDMI, HDMI 5
    HDMI Audio Output
usbstream:CARD=HDMI
    HDA ATI HDMI
    USB Stream Output
sysdefault:CARD=Generic
    HD-Audio Generic, ALCS1200A Analog
    Default Audio Device
front:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    Front output / input
surround21:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Analog
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=Generic,DEV=0
    HD-Audio Generic, ALCS1200A Digital
    IEC958 (S/PDIF) Digital Audio Output
usbstream:CARD=Generic
    HD-Audio Generic
    USB Stream Output
```
Which is (partly) coming from a set of predefined configs:
```
$ ls /usr/share/alsa/alsa.conf.d/

10-samplerate.conf  50-arcam-av-ctl.conf  50-oss.conf       50-pulseaudio.conf  60-upmix.conf     98-maemo.conf       99-pipewire-default.conf
10-speexrate.conf   50-jack.conf          50-pipewire.conf  60-speex.conf       60-vdownmix.conf  98-usb-stream.conf  99-pulseaudio-default.conf
```

Versus just listing the device names (though I'm not sure doing this is correct for cases of audio routing):
```
$ aplay --list-devices

**** List of PLAYBACK Hardware Devices ****
card 0: NVidia [HDA NVidia], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: NVidia [HDA NVidia], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: NVidia [HDA NVidia], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: NVidia [HDA NVidia], device 9: HDMI 3 [HDMI 3]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: NVidia [HDA NVidia], device 10: HDMI 4 [HDMI 4]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: NVidia [HDA NVidia], device 11: HDMI 5 [HDMI 5]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 9: HDMI 3 [HDMI 3]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 10: HDMI 4 [HDMI 4]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA ATI HDMI], device 11: HDMI 5 [HDMI 5]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: Generic [HD-Audio Generic], device 0: ALCS1200A Analog [ALCS1200A Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: Generic [HD-Audio Generic], device 1: ALCS1200A Digital [ALCS1200A Digital]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

**Reply by Tresquel** [reply]
The game isn't crashing tho I did somehow crash bass once but i can't seem to reproduce it.
I'll upload the logs later today.

**Reply by smoogipoo** [reply]
Can you explain what "crash bass" means?

**Reply by Tresquel** [reply]
> Can you explain what "crash bass" means?

the audio library that osu is using

**Reply by smoogipoo** [reply]
I know what BASS is. You said that the game isn't crashing, but what do you mean when you say "crash" BASS? How are you seeing a BASS "crash" which doesn't crash the game?

**Reply by Tresquel** [reply]
I mean it by it just crapping itself and not playing any sound until I restarted the game

**Reply by Tresquel** [reply]
> I'll upload the logs later today.

I was switching thru random outputs and it just kept giving me errors.
At the end I switched to PulseAudioServer or something and it just stopped playing any sound from the game.

[runtime.log](https://github.com/ppy/osu/files/7529054/runtime.log)
[database.log](https://github.com/ppy/osu/files/7529051/database.log)
[network.log](https://github.com/ppy/osu/files/7529052/network.log)
[performance.log](https://github.com/ppy/osu/files/7529053/performance.log)

**Reply by gagahpangeran** [reply]
I can only see some database error from your `database.log`, not sure it is related with your post or not.

**Reply by frenzibyte** [reply]
@Tresquel Were you referring to the following error message when you mentioned it giving you errors?
`2021-11-12 16:36:41 [important]: BASS faulted with error code 46: Busy.`

When BASS fails to initialize an audio device you pick, it'll automatically fall back to using the Default device. And currently switching around between two devices may cause sound to no longer play (see https://github.com/ppy/osu/discussions/15497#discussioncomment-1598854), you can confirm that BASS is still working by changing the current beatmap, or attempt to play a new sample that never played before.

**Reply by Tresquel** [reply]
> you can confirm that BASS is still working by changing the current beatmap, or attempt to play a new sample that never played before.

I tried it and it's true so it's not really a bug?

**Reply by frenzibyte** [reply]
On that part, yes, that would be a duplicate of https://github.com/ppy/osu/discussions/15497#discussioncomment-1598854.

**Reply by smoogipoo** [reply]
Marking as resolved then, nothing to be actioned on here.
